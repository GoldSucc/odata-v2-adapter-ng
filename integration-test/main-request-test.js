"use strict";

const supertest = require("supertest");

const env = require("./_env");
const util = require("../test/_env/util");

let context;
let request;

describe("main-request", () => {
  beforeAll(async () => {
    context = await env("index");
    request = supertest(context.app);
  });

  afterAll(() => {
    env.end(context);
  });

  it("GET $metadata", async () => {
    const response = await util.callRead(request, "/v2/main/$metadata", {
      accept: "application/xml",
    });
    expect(response.body).toBeDefined();
    expect(response.text).toMatchSnapshot();
  });

  it("GET with parameters", async () => {
    const stock = Math.round(new Date().getTime() / 1000);
    await util.callWrite(request, "/v2/main/Header", {
      stock: stock,
      currency: "USD",
    });
    await util.callWrite(request, "/v2/main/Header", {
      stock: 1,
      currency: "EUR",
    });
    const response = await util.callRead(request, `/v2/main/HeaderParameters(STOCK=${stock},CURRENCY='USD')/Set`);
    expect(response.body).toBeDefined();
    expect(response.body.d.results).toHaveLength(1);
    expect(response.body.d.results).toMatchObject([
      {
        __metadata: {
          type: "test.MainService.HeaderParameters",
        },
        country: null,
        createdBy: "anonymous",
        currency: "USD",
        description: null,
        modifiedBy: "anonymous",
        name: null,
        price: null,
        stock,
      },
    ]);
  });

  it("GET request with function 'substringof'", async () => {
    let response = await util.callWrite(request, "/v2/main/Header", {
      name: "Test",
    });
    expect(response.statusCode).toEqual(201);
    const id = response.body.d.ID;
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and substringof('es',name)`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and substringof('ES',name)`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and substringof('XX',name)`);
    expect(response.body.d.results).toHaveLength(0);
  });

  it("GET request with function 'startswith'", async () => {
    let response = await util.callWrite(request, "/v2/main/Header", {
      name: "Test",
    });
    expect(response.statusCode).toEqual(201);
    const id = response.body.d.ID;
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and startswith(name,'Te')`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and startswith(name,'TE')`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and startswith(name,'XX')`);
    expect(response.body.d.results).toHaveLength(0);
  });

  it("GET request with function 'endswith'", async () => {
    let response = await util.callWrite(request, "/v2/main/Header", {
      name: "Test",
    });
    expect(response.statusCode).toEqual(201);
    const id = response.body.d.ID;
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and endswith(name,'st')`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and endswith(name,'ST')`);
    expect(response.body.d.results).toHaveLength(1);
    response = await util.callRead(request, `/v2/main/Header?$filter=ID eq guid'${id}' and endswith(name,'XX')`);
    expect(response.body.d.results).toHaveLength(0);
  });
});
