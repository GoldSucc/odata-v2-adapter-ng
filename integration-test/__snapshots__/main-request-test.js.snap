// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main-requests GET $metadata 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<edmx:Edmx Version=\\"1.0\\" xmlns:edmx=\\"http://schemas.microsoft.com/ado/2007/06/edmx\\" xmlns:m=\\"http://schemas.microsoft.com/ado/2007/08/dataservices/metadata\\" xmlns:sap=\\"http://www.sap.com/Protocols/SAPData\\">
  <edmx:Reference Uri=\\"https://sap.github.io/odata-vocabularies/vocabularies/Common.xml\\" xmlns:edmx=\\"http://docs.oasis-open.org/odata/ns/edmx\\">
    <edmx:Include Alias=\\"Common\\" Namespace=\\"com.sap.vocabularies.Common.v1\\"/>
  </edmx:Reference>
  <edmx:Reference Uri=\\"https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Core.V1.xml\\" xmlns:edmx=\\"http://docs.oasis-open.org/odata/ns/edmx\\">
    <edmx:Include Alias=\\"Core\\" Namespace=\\"Org.OData.Core.V1\\"/>
  </edmx:Reference>
  <edmx:Reference Uri=\\"https://sap.github.io/odata-vocabularies/vocabularies/UI.xml\\" xmlns:edmx=\\"http://docs.oasis-open.org/odata/ns/edmx\\">
    <edmx:Include Alias=\\"UI\\" Namespace=\\"com.sap.vocabularies.UI.v1\\"/>
  </edmx:Reference>
  <edmx:DataServices m:DataServiceVersion=\\"2.0\\">
    <Schema Namespace=\\"test.MainService\\" xmlns=\\"http://schemas.microsoft.com/ado/2008/09/edm\\">
      <EntityContainer Name=\\"EntityContainer\\" m:IsDefaultEntityContainer=\\"true\\">
        <EntitySet Name=\\"Header\\" EntityType=\\"test.MainService.Header\\"/>
        <EntitySet Name=\\"HeaderParameters\\" EntityType=\\"test.MainService.HeaderParametersParameters\\" sap:creatable=\\"false\\" sap:updatable=\\"false\\" sap:deletable=\\"false\\" sap:pageable=\\"false\\"/>
        <EntitySet Name=\\"HeaderParametersSet\\" EntityType=\\"test.MainService.HeaderParametersType\\" sap:creatable=\\"false\\" sap:updatable=\\"false\\" sap:deletable=\\"false\\" sap:addressable=\\"false\\"/>
        <AssociationSet Name=\\"HeaderParametersParameters_Set\\" Association=\\"test.MainService.HeaderParametersParameters_Set\\">
          <End Role=\\"HeaderParametersParameters\\" EntitySet=\\"HeaderParameters\\"/>
          <End Role=\\"HeaderParametersType\\" EntitySet=\\"HeaderParametersSet\\"/>
        </AssociationSet>
      </EntityContainer>
      <EntityType Name=\\"Header\\">
        <Key>
          <PropertyRef Name=\\"ID\\"/>
        </Key>
        <Property Name=\\"ID\\" Type=\\"Edm.Guid\\" Nullable=\\"false\\"/>
        <Property Name=\\"createdAt\\" Type=\\"Edm.DateTimeOffset\\" Precision=\\"7\\"/>
        <Property Name=\\"createdBy\\" Type=\\"Edm.String\\" MaxLength=\\"255\\"/>
        <Property Name=\\"modifiedAt\\" Type=\\"Edm.DateTimeOffset\\" Precision=\\"7\\"/>
        <Property Name=\\"modifiedBy\\" Type=\\"Edm.String\\" MaxLength=\\"255\\"/>
        <Property Name=\\"name\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"description\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"country\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"currency\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"stock\\" Type=\\"Edm.Int32\\"/>
        <Property Name=\\"price\\" Type=\\"Edm.Decimal\\" Scale=\\"2\\" Precision=\\"12\\"/>
      </EntityType>
      <EntityType Name=\\"HeaderParametersParameters\\" sap:semantics=\\"parameters\\">
        <Key>
          <PropertyRef Name=\\"STOCK\\"/>
          <PropertyRef Name=\\"CURRENCY\\"/>
        </Key>
        <Property Name=\\"STOCK\\" Type=\\"Edm.Int32\\" Nullable=\\"false\\"/>
        <Property Name=\\"CURRENCY\\" Type=\\"Edm.String\\" Nullable=\\"false\\"/>
        <NavigationProperty Name=\\"Set\\" Relationship=\\"test.MainService.HeaderParametersParameters_Set\\" FromRole=\\"HeaderParametersParameters\\" ToRole=\\"HeaderParametersType\\"/>
      </EntityType>
      <EntityType Name=\\"HeaderParametersType\\">
        <Key>
          <PropertyRef Name=\\"ID\\"/>
        </Key>
        <Property Name=\\"ID\\" Type=\\"Edm.Guid\\" Nullable=\\"false\\"/>
        <Property Name=\\"createdAt\\" Type=\\"Edm.DateTimeOffset\\" Precision=\\"7\\"/>
        <Property Name=\\"createdBy\\" Type=\\"Edm.String\\" MaxLength=\\"255\\"/>
        <Property Name=\\"modifiedAt\\" Type=\\"Edm.DateTimeOffset\\" Precision=\\"7\\"/>
        <Property Name=\\"modifiedBy\\" Type=\\"Edm.String\\" MaxLength=\\"255\\"/>
        <Property Name=\\"name\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"description\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"country\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"currency\\" Type=\\"Edm.String\\"/>
        <Property Name=\\"stock\\" Type=\\"Edm.Int32\\"/>
        <Property Name=\\"price\\" Type=\\"Edm.Decimal\\" Scale=\\"2\\" Precision=\\"12\\"/>
        <NavigationProperty Name=\\"Parameters\\" Relationship=\\"test.MainService.HeaderParametersParameters_Set\\" FromRole=\\"HeaderParametersType\\" ToRole=\\"HeaderParametersParameters\\"/>
      </EntityType>
      <Association Name=\\"HeaderParametersParameters_Set\\">
        <End Role=\\"HeaderParametersParameters\\" Type=\\"test.MainService.HeaderParametersParameters\\" Multiplicity=\\"0..1\\"/>
        <End Role=\\"HeaderParametersType\\" Type=\\"test.MainService.HeaderParametersType\\" Multiplicity=\\"*\\"/>
      </Association>
      <Annotations Target=\\"test.MainService.Header/createdAt\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Created On\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Immutable\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.Header/createdBy\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Created By\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Description\\" String=\\"User's unique ID\\"/>
        <Annotation Term=\\"Core.Immutable\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.Header/modifiedAt\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Changed On\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.Header/modifiedBy\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Changed By\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Description\\" String=\\"User's unique ID\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.HeaderParametersType/createdAt\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Created On\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Immutable\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.HeaderParametersType/createdBy\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Created By\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Description\\" String=\\"User's unique ID\\"/>
        <Annotation Term=\\"Core.Immutable\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.HeaderParametersType/modifiedAt\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Changed On\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
      <Annotations Target=\\"test.MainService.HeaderParametersType/modifiedBy\\" xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\">
        <Annotation Term=\\"Common.Label\\" String=\\"Changed By\\"/>
        <Annotation Term=\\"Core.Computed\\" Bool=\\"true\\"/>
        <Annotation Term=\\"Core.Description\\" String=\\"User's unique ID\\"/>
        <Annotation Term=\\"UI.HiddenFilter\\" Bool=\\"true\\"/>
      </Annotations>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>"
`;
